sklearn/cluster/dbscan_.py

```diff
@@ -128,15 +128,15 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski', metric_params=None,
     if metric == 'precomputed' and sparse.issparse(X):
         neighborhoods = np.empty(X.shape[0], dtype=object)
         X.sum_duplicates()  # XXX: modifies X's internals in-place
+
+        # set the diagonal to explicit zeros, as a point is its own neighbor
+        X.setdiag(np.zeros(X.shape[0]))
```

Comment:
but you're modifying the input. Is that a good idea?? If the user accidentally put in the wrong sparse matrix, you'd be modifying their data...

Also, this raises SparseEfficiencyWarning in CSR/CSC (and probably should raise it in COO also). We should hide it.

Reply:
Good point! Do you have any idea apart from copying X?
About the warning, I don't get anything, am I missing something?